### PR TITLE
feat(overlay): add basic core of overlay

### PR DESCRIPTION
**Pull request by jelbourn**
R: @hansl @kara 

This is the first step on the journey to overlays. I'll be doing it in as small of pieces as possible for ease of reviewing. The PR covers adding an overlay element to the DOM and attaching a portal to it. Future PRs will cover styling, sizing, positioning, and helper directives.

### CODE REVIEW

File: src/core/overlay/overlay.ts [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+/** Next overlay unique ID. */
+let nextUniqueId = 0;
+
+/** The default state for newly created overlays. */
+let defaultState = new OverlayState();
+
+
+/**
+ * Service to create Overlays. Overlay are dynamically added pieces of floating UI, meant to be
+ * used as low-level a building building block for other components. Dialogs, tooltips, menus,
```
Comment (kara) [inline src/core/overlay/overlay.ts]: A few typos: 
 "Overlay are" --> "Overlay**s** are"
"used as low-level a building building block" --> "used as a low-level building block"

File: src/core/overlay/overlay.ts [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+      private _dynamicComponentLoader: DynamicComponentLoader,
+      private _appViewManager: AppViewManager) {
+  }
+
+  /**
+   * Creates an overlay.
+   * @param state State to apply to the overlay.
+   * @returns A reference to the created overlay.
+   */
+  create(state: OverlayState = defaultState): Promise<OverlayRef> {
```
Comment (kara) [inline src/core/overlay/overlay.ts]: I'm not sure default arguments work with TS2Dart yet. You may have to write this by checking if `isPresent`

File: src/demo-app/overlay/overlay-demo.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  templateUrl: 'demo-app/overlay/overlay-demo.html',
+  styleUrls: ['demo-app/overlay/overlay-demo.css'],
+  directives: [TemplatePortalDirective],
+  providers: [
+    Overlay,
+    provide(OVERLAY_CONTAINER_TOKEN, {useValue: document.body})
+  ]
+})
+export class OverlayDemo {
+  @ViewChildren(TemplatePortalDirective) templatePortals: QueryList<Portal<any>>;
```
Comment (kara) [inline src/demo-app/overlay/overlay-demo.ts]: Looks like there is an extra `>` at the end here?

File: src/core/overlay/overlay.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+  /**
+   * Creates the DOM element for an overlay.
+   * @param state State to apply to the created element.
+   * @returns Promise resolving to the created element.
+   */
+  private _createPaneElement(state: OverlayState): Promise<Element> {
+    var pane = DOM.createElement('div');
+    pane.id  = `md-overlay-${nextUniqueId++}`;
+    pane.classList.add('md-overlay-pane');
```
Comment (kara) [inline src/core/overlay/overlay.ts]: Could we maybe use `DOM.addClass` here instead and pass in the element?

File: src/demo-app/overlay/overlay-demo.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+  constructor(public overlay: Overlay, public elementRef: ElementRef) {
+    BrowserDomAdapter.makeCurrent();
```
Comment (DevVersion) [inline src/demo-app/overlay/overlay-demo.ts]: Just a question: Why are you initializing the BrowserDomAdapater yourself?

Here is my post on Stackoverflow - [Click](http://stackoverflow.com/questions/34686856/angular-2-0-beta-access-dom-adapter-browserdomadapter)
And here's the Angular 2 Issue, which has been created from the stack: [Click](https://github.com/angular/angular/issues/6904)

As you can read in my stackoverflow post, the browser should initialize itself.

Comment (jelbourn) [inline src/demo-app/overlay/overlay-demo.ts]: This isn't actually the `BrowserAdapter` from ng2, it's here in `src/core/platform` copied and pasted from `angular/angular`. It's only necessary for transpiling to Dart right now and will eventually be dropped when https://github.com/angular/angular/issues/7561 is done.

File: src/demo-app/overlay/overlay-demo.ts (hunk 3 above)
Comment (jelbourn) [inline src/demo-app/overlay/overlay-demo.ts]: Not that I can tell.

File: src/core/overlay/overlay.ts (hunk 2 above)
Comment (jelbourn) [inline src/core/overlay/overlay.ts]: They should work- I've seen them used throughout `angular/angular`.

File: src/demo-app/overlay/overlay-demo.ts (hunk 5 above)
Comment (DevVersion) [inline src/demo-app/overlay/overlay-demo.ts]: Oh well, didn't notice the `import`. Thanks.

File: src/demo-app/overlay/overlay-demo.ts (hunk 3 above)
Comment (kara) [inline src/demo-app/overlay/overlay-demo.ts]: You're right; missed the first one before Portal somehow :)

File: src/core/overlay/overlay.ts [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+/** Next overlay unique ID. */
+let nextUniqueId = 0;
+
+/** The default state for newly created overlays. */
+let defaultState = new OverlayState();
+
+
+/**
+ * Service to create Overlays. Overlays are dynamically added pieces of floating UI, meant to be
+ * used as a low-level building building block for other components. Dialogs, tooltips, menus,
```
Comment (kara) [inline src/core/overlay/overlay.ts]: Still has "building" twice?

**Comment by kara** [comment]
LGTM, though still has that one typo if you want to fix

**Comment by hansl** [comment]
@robertmesserle: the e2e is failing on this; https://travis-ci.org/angular/material2/builds/117214613

**Comment by robertmesserle** [comment]
That means it’s working. He didn’t update the screenshot.

**Comment by jelbourn** [comment]
I did update it. Seems like it fails to get the image.

**Comment by robertmesserle** [comment]
@jelbourn Ah, my bad - I was looking on my phone earlier and didn't see the screenshot in the changed files.  Must've missed it.  Anyway, hard to say what is happening, but this is one of the first things that made me want to push for going back to the server idea for screenshots: our method of manually handling `git-lfs` files is definitely hacky and I wouldn't be surprised if that was the cause.

**Comment by robertmesserle** [comment]
Also, after looking at the screenshot URL, it's likely the issue I brought up earlier about PR's that live in other people's repos.  The URL for your screenshot doesn't match the pattern of other screenshots we've dealt with so far, so manually handling the `git-lfs` files might not even really be a valid option.

**Comment by jelbourn** [comment]
@robertmesserle another bullet point for your doc.

**Pull request merged**